common/src/main/java/com/tc/net/protocol/transport/ClientConnectionEstablisher.java

```diff
@@ -396,17 +406,19 @@ private class AsyncReconnect {
      * @param mayInterruptIfRunning interrupt thread
      * @return true if the passed in thread is joined
      */
-    private boolean waitForThread(Thread oldThread, boolean mayInterruptIfRunning) {
+    private boolean waitForConnectionThreadToFinish() {
```

Comment:
I think this method might be clearer structured like this:
```java
private boolean waitForConnectionThreadToFinish() {
  Thread oldThread = getConnectionThread();
  if (oldThread == null) {
    return true;
  } else if (Thread.currentThread() == oldThread) {
    LOGGER.warn("Assertion error, " + oldThread + " cannot join itself", new Exception());
    return false;
  } else {
    oldThread.interrupt();
    try {
       oldThread.join();
      return true;
    } catch (InterruptedException e) {
      LOGGER.info("Got interrupted while waiting for " + oldThread + " to complete");
      Thread.currentThread().interrupt();
      return false;
    }
  }
}
```

Reply:
removed in the latest commit

Reply:
this was poor logic because old thread often == current thread due to the fact that reconnects can be triggered by exceptions and disconnects during the reconnect process.